File: chapter2/scala/src/main/scala/chapter1/mnmcount.scala
// scalastyle:off println

package main.scala.chapter1

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._

/**
  * Usage: MnMcount <mnm_file_dataset>
  */
object MnMcount {
  def main(args: Array[String]) {
    val spark = SparkSession
      .builder
      .appName("MnMCount")
      .getOrCreate()

    if (args.length < 1) {
      print("Usage: MnMcount <mnm_file_dataset>")
      sys.exit(1)
    }
    // get the M&M data set file name
    val mnm_file = args(0)
    // read the file into a Spark DataFrame
    val mnm_df = spark.read.format("csv")
      .option("header", "true")
      .option("inferSchema", "true")
      .load(mnm_file)
    // aggregate count of all colors and groupBy state and color
    // orderBy descending order
    val count_mnm_df = mnm_df.select("State", "Color", "Count")
        .groupBy("State", "Color")
        .agg(count("Count")
        .alias("Total"))
        .orderBy(desc("Total"))
    // show all the resulting aggregation for all the dates and colors
    count_mnm_df.show(60)
    println("Total Rows = %d", (count_mnm_df.count()))
    println()
    //
    // find the aggregate count for California by filtering
    val ca_count_mnm_df = mnm_df.select("*")
      .where(col("State") === "CA")
      .groupBy("State", "Color")
      .agg(count("Count")
      .alias("Total"))
      .orderBy(desc("Total"))
    // show the resulting aggregation for California
    ca_count_mnm_df.show(10)
    //stop the SparkSession
    spark.stop()
  }
}
// scalastyle:on println

File: chapter2/py/src/mnmcount.py
from __future__ import print_function

import sys

from pyspark.sql import SparkSession
from pyspark.sql.functions import count

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: mnmcount <file>", file=sys.stderr)
        sys.exit(-1)

    spark = (SparkSession
        .builder
        .appName("PythonMnMCount")
        .getOrCreate())
    # get the M&M data set file name
    mnm_file = sys.argv[1]
    # read the file into a Spark DataFrame
    mnm_df = (spark.read.format("csv")
        .option("header", "true")
        .option("inferSchema", "true")
        .load(mnm_file))
    mnm_df.show(n=5, truncate=False)
    # aggregate count of all colors and groupBy state and color
    # orderBy descending order
    count_mnm_df = ( mnm_df.select("State", "Color", "Count")
                    .groupBy("State", "Color")
                    .agg(count("Count")
                    .alias("Total"))
                    .orderBy("Total", ascending=False))
    # show all the resulting aggregation for all the dates and colors
    count_mnm_df.show(n=60, truncate=False)
    print("Total Rows = %d" % (count_mnm_df.count()))
    count_mnm_df.explain(True)
    print()
    mnm_df.createOrReplaceTempView("MNM_TABLE_NAME")
    print("Tables and Views:")
    print(spark.catalog.listTables())
    print()
    # Execute the same in SQL
    sql_count_mnn_df = spark.sql("SELECT State, Color, Count, sum(Count) AS Total FROM MNM_TABLE_NAME GROUP BY State, Color, Count ORDER BY Total DESC")
    sql_count_mnn_df.explain(True)
    #
    # find the aggregate count for California by filtering
    ca_count_mnm_df = ( mnm_df.select("*")
                       .where(mnm_df.State == 'CA')
                       .groupBy("State", "Color")
                       .agg(count("Count")
                            .alias("Total"))
                       .orderBy("Total", ascending=False) )

    # show the resulting aggregation for California
    ca_count_mnm_df.show(n=10, truncate=False)

